contracts/Remittance.sol

```diff
@@ -1,16 +1,137 @@
 //SPDX-License-Identifier: MIT
 pragma solidity >= 0.6.0 < 0.7.0;
 
-import "../node_modules/@openzeppelin/contracts/access/Ownable.sol";
-import "../node_modules/@openzeppelin/contracts/utils/Pausable.sol";
+import "./access/Ownable.sol";
+import "./utils/Pausable.sol";
 
-import {SafeMath} from "../node_modules/@openzeppelin/contracts/math/SafeMath.sol";
+import {SafeMath} from "./math/SafeMath.sol";
 
 contract Remittance is Ownable, Pausable {
 
-    using SafeMath for uint;
+    using SafeMath for uint256;
 
+    mapping(bytes32 => RemittanceInstance) public remittances;
 
+    // The 'valid' element provides a flag for checking a bonafide hash key
+    struct RemittanceInstance {
+        address funder;
+        address broker;
+        uint256 fundsOwed;
+        uint256 expiryDate;
+        bool valid;
+    }
+
+    event RemittanceCreated(
+        bytes32 indexed hash,
+        address indexed funder,
+        address indexed broker,
+        uint256 amount,
+        uint256 expiryDate
+    );
+
+    event RemittanceFundsReleased(
+        bytes32 indexed hash,
+        address indexed broker,
+        uint256 amount
+    );
+
+    event RemittanceFundsReclaimed(
+        bytes32 indexed hash,
+        address indexed funder,
+        uint256 amount
+    );
+
+    // The funder can create a remittance
+    // Hash is a sha256 hash of the broker address + recipient password
+    function create(bytes32 recipientPasswordHash, address broker, uint256 expiryDate)
+    external
+    payable
+    whenNotPaused
+    returns(bool success)
+    {
+        require(broker != msg.sender, "The caller cannot be the broker");
+        require(msg.value > 0, "The amount must be greater than 0");
+
+        bytes32 rehash = keccak256ReHash(recipientPasswordHash, broker);
+
+        RemittanceInstance storage remittanceInstance = remittances[rehash];
+
+        remittanceInstance.funder = msg.sender;
+        remittanceInstance.broker = broker;
+        remittanceInstance.fundsOwed = msg.value;
+        remittanceInstance.expiryDate = expiryDate;
+        remittanceInstance.valid = true;
+
+        emit RemittanceCreated(rehash, msg.sender, broker, msg.value, expiryDate);
+
+        return true;
+    }
+
+    // The broker sends the recipient's password to release the funds
+    function release(bytes32 recipientPasswordHash) external whenNotPaused returns(bool success) {
+
+        bytes32 rehash = keccak256ReHash(recipientPasswordHash, address(msg.sender));
+
+        //Retrieve remittance
+        RemittanceInstance storage remittanceInstance = remittances[rehash];
+
+        uint256 fundsOwed = remittanceInstance.fundsOwed;
+
+        require(remittanceInstance.valid, "No remittance found");
+        require(fundsOwed > 0, "No funds available");
+
```

Comment:
What stopped you from also setting `.value = false;`? After all, that could provide some gas refund.

Reply:
> What stopped you from also setting `.value = false;`? After all, that could provide some gas refund.

Did you mean .valid ?

Reply:
Oops yes

Reply:
I created the .valid property so that a call could discern that the supplied mapping key is valid.

If a subsequent call is made with the same hash after a release or reclaim, then I wanted to indicate that the record exists for that hash and revert with a "no funds available" message.

Without that, it would simply check that fundsOwed > 0 but will always return 0 if no record corresponds to the key(hash).

Perhaps I have over-engineered?

Reply:
Oh. Sorry. I just saw your comment below.

Reply:
So you added `.value` so that you could differentiate between these 2 "wrong" situations:

1. `.valid == false && .fundsOwed == 0`, meaning there never were funds in the first place.
2. `.valid == true && .fundsOwed == 0`, meaning there used to be funds, but they have been released.

The only other situation that occurs is:

3. `.valid == true && .fundsOwed > 0`, there are funds to be taken.

The situation that never occurs with your code, although the data structure allows it, is:

4. `.valid == false && .fundsOwed > 0`

So, by disambiguating 1 and 2, you are trying to be nice to the mistaken caller, or to assist the future auditor who does not want to look at the events. At large cost to the careful user.

Reply:
I understand your point.